lib/Doctrine/ODM/MongoDB/Utility/EncryptedFieldsMapGenerator.php

```diff
@@ -0,0 +1,140 @@
+            if ($classMap === []) {
+                continue;
+            }
+
+            $encryptedFieldsMap[$classMetadata->getName()] = $classMap;
+        }
+
+        return $encryptedFieldsMap;
+    }
+
+    /**
+     * Generate the encryption field map from the class metadata.
+     *
+     * @param class-string $className
+     *
+     * @return array<int, array{path: string, bsonType: string, keyId: ?string}>
+     */
+    public function getEncryptedFieldsMapForClass(string $className): array
+    {
+        $classMetadata = $this->classMetadataFactory->getMetadataFor($className);
+
+        return iterator_to_array($this->createEncryptedFieldsMapForClass($classMetadata));
+    }
+
+    /**
+     * @param array<class-string, true> $visitedClasses
+     * @phpstan-param ClassMetadata<T> $classMetadata
+     *
+     * @return Generator<int, array{path: string, bsonType: string, keyId: ?string}>
+     *
+     * @template T of object
+     */
+    private function createEncryptedFieldsMapForClass(
+        ClassMetadata $classMetadata,
+        string $path = '',
+        array $visitedClasses = [],
+    ): Generator {
+        if ($classMetadata->isEncrypted && ! $classMetadata->isEmbeddedDocument) {
+            throw MappingException::rootDocumentCannotBeEncrypted($classMetadata->getName());
+        }
+
+        if (isset($visitedClasses[$classMetadata->getName()])) {
+            // Prevent infinite recursion due to circular references in the metadata
+            return;
+        }
+
+        foreach ($classMetadata->fieldMappings as $mapping) {
+            // @todo support polymorphic types and inheritance?
+            // Add fields recursively
+            if ($mapping['embedded'] ?? false) {
+                $embedMetadata = $this->classMetadataFactory->getMetadataFor($mapping['targetDocument']);
+
+                // When the embedded document class is encrypted, the field is encrypted,
+                // but none of the embedded fields are encrypted separately.
+                if ($embedMetadata->isEncrypted) {
+                    $mapping['encrypt'] ??= []; // @todo get the keyId
+                } elseif (! isset($mapping['encrypt'])) {
+                    yield from $this->createEncryptedFieldsMapForClass(
+                        $embedMetadata,
+                        $path . $mapping['name'] . '.',
+                        $visitedClasses + [$classMetadata->getName() => true],
+                    );
+                }
+            }
+
+            if (! isset($mapping['encrypt'])) {
+                continue;
+            }
+
+            $field = [
+                'path' => $path . $mapping['name'],
+                'bsonType' => match ($mapping['type']) {
+                    ClassMetadata::ONE, Type::HASH => 'object',
+                    ClassMetadata::MANY, Type::COLLECTION => 'array',
+                    Type::INT, Type::INTEGER => 'int',
```

Comment:
The [BSON type](https://www.mongodb.com/docs/manual/reference/bson-types/#std-label-bson-types) mappings for encryption differentiates between 32-bit and 64-bit integers (`int` and `long`, respectively). This could be problematic, as the match here is going to limit all integer-mapped fields to 32-bit. Defaulting to 64-bit isn't necessarily a workaround, as the server is going to expect matching types for any equality queries.

I believe this is being tracked by [SERVER-75897](https://jira.mongodb.org/browse/SERVER-75897).

Reply:
So, do you suggest we should change it to `long`?
```suggestion
                    Type::INT, Type::INTEGER => 'long',
```

Reply:
According to the [CSFLE Encryption Schemas: `encrypt` docs](https://www.mongodb.com/docs/manual/core/csfle/reference/encryption-schemas/#mongodb-autoencryptkeyword-autoencryptkeyword.encrypt), we can specify an array of types for `bsonType`. Can we use `['int', 'long']` here to remain flexible?

I'll also share this question in the CSFLE Slack for a sanity check.

Reply:
@kevinalbs replied in Slack:

> Notably: though CSFLE schema (JSON Schema) supports multiple BSON types, a QE schema (`encryptedFields`) only supports a single valued bsonType.

So I don't think we have a user-friendly solution here unless we prefer "long" (since it can contain _any_ integer value) and have ODM _always_ encodes the encrypted field as a `MongoDB\BSON\Int64` to ensure even values in the 32-bit range are stored as int64 and thus satisfy the QE `bsonType` restriction.

Reply:
When sending a int64 to a field encoded with the int type, we get this error:
```
mongocryptd error: Cannot encrypt element of type long because schema requires that type is one of: [ int ]
```

Also, I cannot set multiple types:

```
mongocryptd error: BSON field 'create.encryptedFields.fields.bsonType' is the wrong type 'array', expected type 'string'                                                            
```

If I change the bson type to `long`, then it doesn't accept an `int`.

It seems that its a limitation until we add the `Int64` type in the ODM.

Reply:
New Int64 type: https://github.com/doctrine/mongodb-odm/pull/2805

Reply:
Well, that was a fun little detour ⚠️